### PR TITLE
source-amazon-seller-partner: allow to use IAM user arn or IAM role arn

### LINKED ISSUE #5942: [kubernetes/helm] Allow authentication to AWS using roles
## Tell us about the problem you're trying to solve

Current version of Airbyte I tested (v0.29.15-alpha) requests providing access key Id and secret access key in all cases where authentication for AWS is required. 

## Describe the solution you’d like

It would be good to allow using roles instead i.e provide the option to use either/whichever works for the user.

## Describe the alternative you’ve considered or used
N/A

## Additional context
N/A

### PR DESCRIPTION
**Pull request by pecalleja**
## What
* The amazon sp-api integration allows to use IAM user entity or IAM role entity.  [amazon documentation](https://developer-docs.amazon.com/sp-api/docs/creating-and-configuring-iam-policies-and-entities#step-5-add-an-aws-security-token-service-aws-sts-policy-to-your-iam-user) 
if we use a user based arn we need to use another function from the boto3 client to get the required credentials role

* This will allow to integrate with airbyte apps configured with user arn entity in the amazon seller central app.

## How
just use the required function for differents  IAM entities 

## 🚨 User Impact 🚨
None

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [X] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [X] Secrets in the connector's spec are annotated with `airbyte_secret`
- [X] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [X] Code reviews completed
- [X] Documentation updated
    - [X] Connector's `README.md`
    - [X] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [X] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [X] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Comment by alafanechere** [comment]
Thank you @pecalleja for this contribution. I'm running the tests and will get back for a first review once they pass.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py [hunk 1]
```diff
@@ -121,6 +120,23 @@ def _get_stream_kwargs(self, config: ConnectorConfig) -> Mapping[str, Any]:
...
+            role = self._assume_role(boto3_client, config.role_arn)
+        else:
+            raise ValueError("Invalid User/Role Arn")
+        return role
+
+    def _get_session_token(self, boto3_client):
+        return boto3_client.get_session_token()
+
+    def _assume_role(self, boto3_client, role_arn: str):
+        return boto3_client.assume_role(RoleArn=role_arn, RoleSessionName="guid")
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:138]: I would suggest to not creating dedicated one-line functions and calling boto client's methods directly from `get_role`

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py [hunk 2]
```diff
@@ -121,6 +120,23 @@ def _get_stream_kwargs(self, config: ConnectorConfig) -> Mapping[str, Any]:
         }
         return stream_kwargs
 
+    def get_role(self, config: ConnectorConfig) -> dict:
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:123]:
```suggestion
    def get_sts_credentials(self, config: ConnectorConfig) -> dict:
```

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py [hunk 3]
```diff
@@ -92,8 +92,7 @@ class SourceAmazonSellerPartner(AbstractSource):
     def _get_stream_kwargs(self, config: ConnectorConfig) -> Mapping[str, Any]:
         endpoint, marketplace_id, region = get_marketplaces(config.aws_environment)[config.region]
 
-        boto3_client = boto3.client("sts", aws_access_key_id=config.aws_access_key, aws_secret_access_key=config.aws_secret_key)
-        role = boto3_client.assume_role(RoleArn=config.role_arn, RoleSessionName="guid")
+        role = self.get_role(config)
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:95]:
```suggestion
        sts_crendentials = self.get_role(config)
```

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py [hunk 4]
```diff
@@ -121,6 +120,23 @@ def _get_stream_kwargs(self, config: ConnectorConfig) -> Mapping[str, Any]:
         }
         return stream_kwargs
 
+    def get_role(self, config: ConnectorConfig) -> dict:
+        boto3_client = boto3.client("sts", aws_access_key_id=config.aws_access_key, aws_secret_access_key=config.aws_secret_key)
+        *_, arn_resource = config.role_arn.split(":")
+        if arn_resource.startswith("user"):
+            role = self._get_session_token(boto3_client)
+        elif arn_resource.startswith("role"):
+            role = self._assume_role(boto3_client, config.role_arn)
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:129]:
```suggestion
            role = self._assume_role(boto3_client, config.role_arn)
```
```suggestion
            sts_credentials = self._assume_role(boto3_client, config.role_arn).get("Credentials)
```

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py [hunk 5]
```diff
@@ -121,6 +120,23 @@ def _get_stream_kwargs(self, config: ConnectorConfig) -> Mapping[str, Any]:
...
+        else:
+            raise ValueError("Invalid User/Role Arn")
+        return role
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:132]:
```suggestion
        return sts_credentials
```

**Review by alafanechere — CHANGES_REQUESTED** [review]
Minor suggestions

**Comment by pecalleja** [comment]
do you want I include more unit tests in this PR to fully cover the connector source code?

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pecalleja) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:123]: sure !!!

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py [hunk 6]
```diff
@@ -121,6 +120,23 @@ def _get_stream_kwargs(self, config: ConnectorConfig) -> Mapping[str, Any]:
         }
         return stream_kwargs
 
+    def get_role(self, config: ConnectorConfig) -> dict:
+        boto3_client = boto3.client("sts", aws_access_key_id=config.aws_access_key, aws_secret_access_key=config.aws_secret_key)
+        *_, arn_resource = config.role_arn.split(":")
```
[the remark this answers is not in the document]
Comment (pecalleja) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:125]: good question... I haven't see it before a function like that in the boto3 library

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py (hunk 1 above)
Comment (pecalleja) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:138]: agree

**Comment by alafanechere** [comment]
> do you want I include more unit tests in this PR to fully cover the connector source code?

I'd prefer you only add tests for the code you add. Feel free to open another PR if you want to increase the coverage of other functions.

File: airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/test_source.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+def test_spec(connector_source):
+    assert isinstance(connector_source.spec(), ConnectorSpecification)
+
+
+def test_streams(connector_source, connector_config, mock_boto_client):
+    for stream in connector_source.streams(connector_config):
+        assert isinstance(stream, Stream)
+
+
+def test_stream_user_role(connector_source, connector_config, mock_boto_client):
```
[the remark this answers is not in the document]
Comment (pecalleja) [reply airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/test_source.py:58]: ammm.. I'm not sure if this test is clear enough. I can sumbit a different approach if you want.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:132]: https://docs.aws.amazon.com/cdk/api/v1/python/aws_cdk.core/Arn.html
I don't know if aws_cdk.core is available in boto 🤔 
It's not really important, only sugar

File: airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/test_source.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/test_source.py]: With this snippet, I wanted to suggest writing a test that asserts that two different functions are called if the arn is for a user or a role.

**Review by alafanechere — CHANGES_REQUESTED** [review]
Thank you @pecalleja, we are close to having a mergeable PR 😄 
I need to ask you some extra work to make your change available in the published connector: 
* Bump the connector version in 
  *  `Dockerfile`
  *  `airbyte-config/init/src/main/resources/seed/source_definitions.yaml`
  * `airbyte-config/init/src/main/resources/seed/source_specs.yaml` 

Update the changelog in `docs/integrations/sources/amazon-seller-partner.md`

I can't do these changes myself as I've no permission to push on your branch.

**Comment by alafanechere** [comment]
Could you also please merge master into your branch? This is probably the reason the latest test fail.

**Review by pecalleja — COMMENTED** [review]
requested changes implemented 👍. Thank you.

File: airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/test_source.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+
+@pytest.mark.parametrize(
+    "arn",
+    ("arn:aws:iam::123456789098:user/some-user", "arn:aws:iam::123456789098:role/some-role")
+)
+def test_stream_with_good_iam_arn_value(connector_source, connector_config, mock_boto_client, arn):
+    connector_config.role_arn = arn
+    result = connector_source.get_sts_credentials(connector_config)
+    assert "Credentials" in result
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/test_source.py:65]: I still find this test a bit weak as it's not checking that something different happens for role and user. Please assert specific function calls on a mocked client

Comment (pecalleja) [reply airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/test_source.py:65]: sorry I don't understand what you mean 😔, can you help me with this please?

Comment (alafanechere) [reply airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/test_source.py:65]:
```

from source_amazon_seller_partner import source

@pytest.mark.parametrize(
    "arn",
    ("arn:aws:iam::123456789098:user/some-user", "arn:aws:iam::123456789098:role/some-role")
)
def test_stream_with_good_iam_arn_value(mocker, connector_source, connector_config, arn):
    mocker.path.object(source, "boto3.client")
    mock_boto_client = source.boto3.client.return_value
    connector_config.role_arn = arn
    result = connector_source.get_sts_credentials(connector_config)
    assert "Credentials" in result
    if "user" in arn:
        mock_boto_client.get_session_token.assert_called_once()
    if "role" in arn:
        mock_boto_client.assume_role.assert_called_once_with(RoleArn=arn, RoleSessionName="guide")
```

Comment (pecalleja) [reply airbyte-integrations/connectors/source-amazon-seller-partner/unit_tests/test_source.py:66]: Thanks 🙏 . I learned things with this.

**Review by pecalleja — COMMENTED** [review]
LGTM

**Comment by alafanechere** [comment]
/publish connector=connectors/source-amazon-seller-partner auto-bump-version=false

**Review by alafanechere — APPROVED** [review]
Thank you @pecalleja for your changes!

**Pull request merged**